Tests: More use of base.d class functions for test code

hewison-chris opened the pull request:
Make more use of `generateBlocks`, `getSpendables` and `postAndEnsureTxInPool` to reduce the lines of code in the tests and make it more maintainable and reliable.

First 3 commits are from PR https://github.com/bosagora/agora/pull/2561

hewison-chris committed:
Tests: Simplify `test.EnrollDifferentUTXO`

Making use of `generateBlocks`, `postAndEnsureTxInPool`, `freezeUtxo`
and `getSpendables` from base class.

hewison-chris committed:
Tests: fix flakey fee test

Use the `getSpendables` function which checks for `utxo` not yet in pool
to make `transaction` for `frozen outputs` to use in `outsider`
`enrollments`. Also use `network.postAndEnsureTxInPool` to make test
clearer and more robust.

hewison-chris committed:
Tests: use `spendables` from `utxo_set` in `addBlock`

It is more reliable to use utxo_set of the first node to find utxo that
are not already in a transaction as input in the transaction pool.
Also to make code cleaner the `postAndEnsureTxInPool` function replaces
the manual `postTransaction` and `hasTransactionHash` in this function.

hewison-chris committed:
Tests: Simplify `test.Ledger`

Making use of `postAndEnsureTxInPool` from base class.

hewison-chris committed:
Tests: Simplify `test.ValidatorRecurringEnrollment`

Making use of `generateBlocks` from base class.

hewison-chris committed:
Tests: Simplify `test.ValidatorCleanRestart`

Making use of `generateBlocks`, `postAndEnsureTxInPool` and
`getSpendables` from base class. This also fixes the test so it can be
enabled to run again.

hewison-chris committed:
Tests: Simplify `test.RestoreSlashingInfo`

Add `freezeUtxo` to base test class and make use of it and
`postAndEnsureTxInPool` and `getSpendables`.

linked0 commented on source/agora/test/Base.d:
```diff
@@ -1443,7 +1443,7 @@ public interface TestAPI : API
     ***************************************************************************/
 
-    public UTXOPair[] getUTXOs (Amount minimum);
+    public UTXOPair[] getSpendables (Amount minimum);
```
Change doc, please.

hewison-chris replied:
I updated it in PR https://github.com/bosagora/agora/pull/2561 which is the first 3 commits of this PR.

linked0 commented on source/agora/test/Base.d:
```diff
@@ -1443,7 +1443,7 @@ public interface TestAPI : API
+    public UTXOPair[] getSpendables (Amount minimum);
 
     ///
     public UTXOPair[] getUTXOs (PublicKey owner);
```
It needs doc

hewison-chris replied:
Also updated in other PR

linked0 commented on source/agora/test/Fee.d:
```diff
@@ -35,16 +37,16 @@ unittest
+        network.postAndEnsureTxInPool(
+            TxBuilder(WK.Keys.AAA.address)
+            .attach(utxo_pairs.map!(p => tuple(p.utxo.output, p.hash)))
+            .deduct(1.coins).sign());
```
Out of curiosity, why do we need `deduct(1.coins)` although the test succeeds without it?

hewison-chris replied:
It just sets the fee to 1 coin. Without it the TxBuilder will calculate the minimum fees required and that will be deducted to create the fees. It could be left out.

linked0 commented on source/agora/test/Fee.d:
```diff
@@ -179,16 +181,16 @@ unittest
+        network.postAndEnsureTxInPool(
+            TxBuilder(WK.Keys.AAA.address)
+            .attach(utxo_pairs.map!(p => tuple(p.utxo.output, p.hash)))
+            .deduct(1.coins).sign());
```
Ditto

hewison-chris replied:
Same.

linked0 commented on source/agora/test/ValidatorCleanRestart.d:
```diff
@@ -59,24 +61,29 @@ version(none) unittest
 
     // wait for other nodes to get to same block height
-    network.assertSameBlocks(Height(GenesisValidatorCycle - 1));
+    set_b.enumerate.each!((idx, node) =>
```
Why do we change this although the `assertSameBlocks` works?

hewison-chris replied:
We want to wait for outsiders to get to the same height which is what the change does. Don't forget the test was disabled before these changes.

linked0 commented on source/agora/test/ValidatorCleanRestart.d:
```diff
@@ -59,24 +61,29 @@ version(none) unittest
 
     // Block 20, After this the Genesis block enrolled validators will be expired.
-    network.generateBlocks(Height(GenesisValidatorCycle));
+    network.generateBlocks(iota(GenesisValidators, GenesisValidators + conf.outsider_validators),
```
Why do we change this?

hewison-chris replied:
It needs to generate the blocks using the `outsiders`.

linked0 reviewed (approved):
Really good. Some comments.
It has CI failures though.

hewison-chris committed:
Tests: Only sleep one node in `MultiRoundConsensus`

To make it more reliable as sometimes it failed with the message:
`The validator's round number is 1. Expected: above 2`.

hewison-chris committed:
Tests: `FutureEnrollment` use `getSpendables`

This ensures that the utxo used for creating the frozen utxo for
enrolling the outsider is not being spent already.

hewison-chris committed:
Tests: Ensure required signature count in `BlockRewards`

In the last test in this file we could not use quorum_threshold of 100%
which is how the other tests ensure that all signatures are present. For
this test the signature count is different for even block heights. To
prevent ocasional failures the test now waits after each block for the
desired signature count so that the calculations are correct for the
expected rewards and fees.

hewison-chris commented:
Closing this as most commits were extracted into PR https://github.com/bosagora/agora/pull/2573
